### ISSUE TITLE
[BUG]: CopyFilesOverSSH fails to copy file to a Windows machine using a non-UNC absolute path

### ISSUE THREAD

**Issue by manokara** [microsoft/azure-pipelines-tasks#18386]
### Task name

CopyFilesOverSSH

### Task version

0

### Environment type (Please select at least one enviroment where you face this issue)

- [ ] Self-Hosted
- [X] Microsoft Hosted
- [ ] VMSS Pool
- [ ] Container

### Azure DevOps Server type

dev.azure.com (formerly visualstudio.com)

### Operation system

Ubuntu 22.04.2

### Task log

```shell
##[debug]Evaluating condition for step: 'Send build archive to server'
##[debug]Evaluating: SucceededNode()
##[debug]Evaluating SucceededNode:
##[debug]=> True
##[debug]check path : /home/vsts/work/1/a
##[debug]targetFolder=C:/LIVE/MYAPP
##[debug]cleanTargetFolder=false
##[debug]overwrite=true
##[debug]failOnEmptySource=true
##[debug]flattenFolders=false
Setting up SSH service connection to remote host XXX.XX.X.XXX.
##[debug]include content pattern: myapp-production-191.7z
##[debug]findPath: '/home/vsts/work/1/a'
##[debug]findOptions.allowBrokenSymbolicLinks: 'false'
##[debug]findOptions.followSpecifiedSymbolicLink: 'true'
##[debug]findOptions.followSymbolicLinks: 'true'
##[debug]findOptions.skipMissingFiles: 'false'
##[debug]  /home/vsts/work/1/a (directory)
##[debug]  /home/vsts/work/1/a/myapp-production-191.7z (file)
##[debug]2 results
##[debug]counted 1 files in the source tree
##[debug]Include matching /home/vsts/work/1/a/myapp-production-191.7z
##[debug]Include matched 1 files
##[debug]Number of files to copy = 1
##[debug]filesToCopy = /home/vsts/work/1/a/myapp-production-191.7z
Found 1 files to copy to the remote machine.
##[debug]fileToCopy = /home/vsts/work/1/a/cmyapp-production-191.7z
##[debug]relativePath = myapp-production-191.7z
Copying file /home/vsts/work/1/a/myapp-production-191.7z to ./C:/LIVE/MYAPP/myapp-production-191.7z on remote machine.
##[debug]Upload /home/vsts/work/1/a/myapp-production-191.7z to ./C:/LIVE/MYAPP/myapp-production-191.7z on remote machine.
##[error]Failed to copy /home/vsts/work/1/a/myapp-production-191.7z. Unable to create target folder ./C:/LIVE/MYAPP.
##[debug]Processed: ##vso[task.issue type=error;]Failed to copy /home/vsts/work/1/a/myapp-production-191.7z. Unable to create target folder ./C:/LIVE/MYAPP.
Completed copying 1 files to the remote machine.
##[debug]task result: Failed
##[error]Failed to copy 1 files
##[debug]Processed: ##vso[task.issue type=error;]Failed to copy 1 files
##[debug]Processed: ##vso[task.complete result=Failed;]Failed to copy 1 files
##[debug]Closing the client connection
##[debug]sftpClient: Ignoring error diconnecting: Error: read ECONNRESET
##[debug]Task successfully accomplished
##[debug]sshClient: Ignoring error diconnecting: Error: read ECONNRESET
Finishing: Send build archive to server
```

### Relevant log output

```shell
Copying file /home/vsts/work/1/a/myapp-production-191.7z to ./C:/LIVE/MYAPP/myapp-production-191.7z on remote machine.
##[debug]Upload /home/vsts/work/1/a/myapp-production-191.7z to ./C:/LIVE/MYAPP/myapp-production-191.7z on remote machine.
##[error]Failed to copy /home/vsts/work/1/a/myapp-production-191.7z. Unable to create target folder ./C:/LIVE/MYAPP.
##[debug]Processed: ##vso[task.issue type=error;]Failed to copy /home/vsts/work/1/a/myapp-production-191.7z. Unable to create target folder ./C:/LIVE/MYAPP.
```

### Aditional info

Basically I think the problem is here:

https://github.com/microsoft/azure-pipelines-tasks/blob/6de2704788d5b29fcf6436ef3875095d91eaa9fb/Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts#L209-L213

If the agent is being run on Ubuntu, `path.isAbsolute()` says that `C:/LIVE/MYAPP` is a relative path. Back when I was using classic release pipelines on this project it worked fine, but now that I I turned it into a deployment stage in YAML this happened. Perhaps because back then it was using a Windows agent to run the release pipeline (and it always used it? even though I never specified it), but now the agent is always running the pipeline on Ubuntu, which is not a problem since I'm deploying a .NET 6 ASP.NET application and scripting with Powershell, so it should work fine in any of the supported systems.

Fortunately, it can be worked around by using `/LIVE/MYAPP` or `\\?\C:\LIVE\MYAPP` as the target folder or forcing the pipeline to run on Windows, but I believe this to be a bug. The task should use `isWindowsOnTarget` to make sure that the target path isn't wrongly recognized as a relative one in case the agent is not running on Windows.

**Comment by ivanduplenskikh** [comment]
Hi @manokara thanks for reporting! We are working on more prioritized issues at the moment, but will get back to this one soon.

**Comment by a-poluyanov** [comment]
Same thing, I'm facing the same problem.
I set the path and indicate that this is a windows machine
<img width="589" alt="Снимок экрана 2024-03-21 в 17 32 01" src="https://github.com/microsoft/azure-pipelines-tasks/assets/81773011/9e2218d8-bd87-42c5-a2f1-21dc1d046d7b">

but the pipeline itself was still set in the wrong direction
<img width="505" alt="Снимок экрана 2024-03-21 в 17 32 31" src="https://github.com/microsoft/azure-pipelines-tasks/assets/81773011/adc49bd2-e539-4b97-adf4-e02eccaf5342">

**Comment by a-poluyanov** [comment]
@manokara
I had the same problem as you. I copied my artifacts from a linux agent to a windows machine. I solved my problem like this:

It’s logical to assume that if files are copied via ssh, then we get to the windows machine using the user, in my case it was Administrator, by default when connecting we end up in his home directory
`C:\Users\Administrator`

I needed to copy files to
`C:\inetpub\wwwroot`

In Azure devops, in my pipeline in the `Copy files over SSH` task, I entered the following `Target folder`:

```
steps:
- task: CopyFilesOverSSH@0
  displayName: 'Securely copy files to the remote machine'
  inputs:
    sshEndpoint: 'your-endpoint'
    sourceFolder: '$(Build.ArtifactStagingDirectory)/dist/'
    targetFolder: /../../inetpub/wwwroot/
    isWindowsOnTarget: true
```
    
it turns out that I went down two folders back and went to the `inetpub/wwwroot` I needed

I hope this helps someone until the developer of this task fixes this problem. This is by no means a final solution and does not mean that we need to close this discussion, the developer still has problems with this task, and we are waiting for a competent and simple fix.
